libraries/custom_appbase/include/eosio/chain/application.hpp

```diff
@@ -30,14 +30,38 @@ class priority_queue_executor {
    static constexpr uint16_t minimum_runtime_ms = 3;
 
    // inform how many read_threads will be calling read_only/read_exclusive queues
-   // Currently only used to assert if exec_queue::read_exclusive is used without any read threads
+   // expected to only be called at program startup, not thread safe, not safe to call after startup
    void init_read_threads(size_t num_read_threads) {
       pri_queue_.init_read_threads(num_read_threads);
    }
 
+   // not thread safe, see init_read_threads comment
+   size_t get_read_threads() const {
+      return pri_queue_.get_read_threads();
+   }
+
+   // not thread safe, only call at program startup from the main thread (thread that calls app().exec()
+   void init_main_thread_id() {
+      main_thread_id_ = std::this_thread::get_id();
+   }
+
+   // not thread safe, but as long as set_main_thread_id() only called at program startup from main thread before
+   // other threads are created then this will be safe to access.
+   std::thread::id get_main_thread_id() const {
+      assert(main_thread_id_ != std::thread::id());
```

Comment:
Why not initialize this in the member, so it is initialized when the executor is constructed?
`std::thread::id                    main_thread_id_ { std::thread::id()) };`

Reply:
I was not thinking it would always be created on the main thread. But in practice I guess it would be.

Reply:
Actually the tests create it on a diff thread. I can re-work those tests.

Reply:
I assumed the `Application` was always created on the main thread, but if that's not the case you can ignore my comment.

Reply:
In the nomial use-case it is, it just isn't for our tests, but I should be able to fix the tests.